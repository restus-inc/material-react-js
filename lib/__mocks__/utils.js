/* global jest */
const utils = jest.genMockFromModule('../utils');

const newId = () => 'a0123456';
utils.newId = newId;

export default utils;
export { newId };
