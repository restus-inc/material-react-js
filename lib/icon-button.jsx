/**
 * @license
 * Copyright 2020 Restus Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { useRef, useCallback } from 'react';
import { MDCRipple } from '@material/ripple';
import { useMDCComponent } from './hoocks';

const NON_NATIVE_PROPS = [
  'className', 'icon', 'ref', 'disablesMdcInstance', 'mdcRippleRef',
];

const generateRootClassName = (props) => {
  const rootClassNames = ['mdc-icon-button'];
  if (props.className) {
    rootClassNames.push(props.className);
  }
  return rootClassNames.join(' ');
};

/**
 * [MDCIconButton component]{@link https://github.com/material-components/material-components-web/tree/master/packages/mdc-icon-button#readme}
 * implemented by react component.
 * @function IconButton
 * @param {Object} props Attributes other than followings are passed to the `button`
 * element of React as is.
 * @param {string} [props.className] The class name that is added to the root element.
 * @param {string} [props.icon] The inner text of the icon element if adding the icon.
 * @param {boolean} [props.disablesMdcInstance] Specifies `true` if you do not want to
 * instantiate MDCRipple Component. Default to `false`.
 * @param {React.MutableRefObject} [props.mdcRippleRef] MutableRefObject which bind an
 * MDCRipple instance to. The instance is not bind if `props.disablesMdcInstance` is `true`.
 * @returns {DetailedReactHTMLElement}
 * @example
 * import { IconButton } from 'material-react-js';
 *
 * function MyFavoriteButton(props) {
 *   switch (props.type) {
 *     case 'material':
 *       return <IconButton className="material-icons" icon="favorite"/>;
 *       // The following code also returns the same output
 *       // return <IconButton className="material-icons">favorite</IconButton>;
 *     case 'image':
 *       return <IconButton><img src="./img/favorite.png" alt="favorite"/></IconButton>;
 *     case 'svg':
 *       return (
 *         <IconButton>
 *           <svg viewBox="0 0 32 32">...</svg>
 *         </IconButton>
 *       );
 *     default:
 *       throw new Error('unknown icon type');
 *   }
 * }
 * @exports material-react-js
 */
export default function IconButton(props) {
  const rootElementRef = useRef();
  useMDCComponent(
    MDCRipple,
    rootElementRef,
    props.mdcRippleRef,
    props.disablesMdcInstance,
    useCallback((mdcRipple) => {
      mdcRipple.unbounded = true; // eslint-disable-line no-param-reassign
    }, []),
  );

  const buttonProps = Object.entries(props).reduce((newProps, [key, value]) => (
    !NON_NATIVE_PROPS.includes(key) ? Object.assign(newProps, { [key]: value }) : newProps
  ), { className: generateRootClassName(props), ref: rootElementRef });

  return <button {...buttonProps}>{props.children || props.icon}</button>;
}
