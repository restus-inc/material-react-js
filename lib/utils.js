/**
 * Utilities for Material React.
 * @module material-react/lib/utils
 */

/**
 * Generates a rundom ID. The generated ID starts with basic latin child alphabet, and
 * subsequent characters consist of child alphanumeric character. (i.e. it matches
 * `/[a-z][a-z0-9]{digit-1}/`)
 * @param {number} [digit=8] The number of digit of the generated ID
 * @returns {string} The ID generated
 */
export function newId(digit = 8) {
  return (
    (Math.floor(Math.random() * 26) + 10) * (36 ** (digit - 1))
    + Math.floor(Math.random() * (36 ** (digit - 1)))
  ).toString(36);
}

/**
 * Converts the event name to event handler name.
 * @param {string} eventName The event name that is converted to the handler name
 * @return {string} The event handler name converted
 */
export function handlerName(eventName) {
  return `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
}
