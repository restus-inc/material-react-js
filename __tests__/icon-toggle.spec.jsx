/**
 * @license
 * Copyright 2020 Restus Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* global jest, describe, afterEach, it, expect */
import 'regenerator-runtime/runtime';
import React, { useEffect, useRef } from 'react';
import { MDCIconButtonToggle } from '@material/icon-button';
import { MDCRipple } from '@material/ripple';
import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { htmlOfRendering } from './utils';

import IconToggle from '../lib/icon-toggle';

describe('IconToggle component', () => {
  afterEach(async () => {
    await cleanup();
  });

  it('supports material icon toggle buttons', () => {
    expect(htmlOfRendering(
      <IconToggle iconClassName="material-icons" onIcon="visibility" offIcon="visibility_off" label="visibility"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <IconToggle iconClassName="material-icons" onIcon="visibility" offIcon="visibility_off" label={['off', 'on']}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <IconToggle isOnState={true} className="foo" label="visibility">
        <IconToggle.OffIcon className="material-icons">visibility_off</IconToggle.OffIcon>
        <IconToggle.OnIcon className="material-icons">visibility</IconToggle.OnIcon>
      </IconToggle>,
    )).resolves.toMatchSnapshot();
  });

  it('supports image icon toggle buttons', () => {
    expect(htmlOfRendering(
      <IconToggle className="foo" label={['off', 'on']}>
        <IconToggle.OffImage src="./img/visibility_off.png" alt="visibility_off"/>
        <IconToggle.OnImage className="bar" src="./img/visibility.png" alt="visibility"/>
      </IconToggle>,
    )).resolves.toMatchSnapshot();
  });

  it('supports SVG icon toggle buttons', () => {
    expect(htmlOfRendering(
      <IconToggle isOnState={true} className="foo" label={['off', 'on']}>
        <IconToggle.OffSVG className="bar" viewBox="0 0 24 24">
          <circle cx="12" cy="12" r="11" strokeWidth="2"/>
        </IconToggle.OffSVG>
        <IconToggle.OnSVG className="baz" viewBox="0 0 24 24">
          <line x1="1" y1="1" x2="23" y2="23" strokeWidth="2"/>
        </IconToggle.OnSVG>
      </IconToggle>,
    )).resolves.toMatchSnapshot();
  });

  it('can provide an MDCIconButtonToggle instance', async () => {
    let mdcIconButtonToggleComponent;
    let mdcRippleComponent;
    function MyIconButtonToggle() {
      const mdcIconButtonToggleRef = useRef();
      const mdcRippleRef = useRef();
      useEffect(() => {
        mdcIconButtonToggleComponent = mdcIconButtonToggleRef.current;
        mdcRippleComponent = mdcRippleRef.current;
      });
      return <IconToggle iconClassName="material-icons"
                         onIcon="visibility" offIcon="visibility_off"
                         mdcIconButtonToggleRef={mdcIconButtonToggleRef}
                         mdcRippleRef={mdcRippleRef}/>;
    }
    render(<MyIconButtonToggle/>);
    expect(mdcIconButtonToggleComponent).toBeInstanceOf(MDCIconButtonToggle);
    expect(mdcRippleComponent).toBeInstanceOf(MDCRipple);
  });

  it('fires onChange event when the icon is toggled', async () => {
    let eventDetail = null;
    const onChange = jest.fn((event) => {
      eventDetail = event.detail;
    });

    const { getByTestId } = render(
      <IconToggle iconClassName="material-icons" onIcon="visibility" offIcon="visibility_off" label="visibility" onChange={onChange} data-testid="toggle"/>,
    );
    const toggleButton = getByTestId('toggle');
    expect(toggleButton.classList).not.toContain('mdc-icon-button--on');
    userEvent.click(getByTestId('toggle'));
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(eventDetail).toEqual({ isOn: true });
    expect(toggleButton.classList).toContain('mdc-icon-button--on');
    userEvent.click(getByTestId('toggle'));
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(eventDetail).toEqual({ isOn: false });
    expect(toggleButton.classList).not.toContain('mdc-icon-button--on');
  });
});
