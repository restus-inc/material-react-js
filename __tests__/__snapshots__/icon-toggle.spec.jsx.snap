// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconToggle component supports SVG icon toggle buttons 1`] = `
"<button
  aria-label=\\"on\\"
  class=\\"mdc-icon-button mdc-icon-button--on foo mdc-ripple-upgraded--unbounded\\"
  data-aria-label-off=\\"off\\"
  data-aria-label-on=\\"on\\"
>
  <svg
    class=\\"mdc-icon-button__icon bar\\"
    viewBox=\\"0 0 24 24\\"
  >
    <circle
      cx=\\"12\\"
      cy=\\"12\\"
      r=\\"11\\"
      stroke-width=\\"2\\"
    />
  </svg>
  <svg
    class=\\"mdc-icon-button__icon mdc-icon-button__icon--on baz\\"
    viewBox=\\"0 0 24 24\\"
  >
    <line
      stroke-width=\\"2\\"
      x1=\\"1\\"
      x2=\\"23\\"
      y1=\\"1\\"
      y2=\\"23\\"
    />
  </svg>
</button>"
`;

exports[`IconToggle component supports image icon toggle buttons 1`] = `
"<button
  aria-label=\\"off\\"
  class=\\"mdc-icon-button foo mdc-ripple-upgraded--unbounded\\"
  data-aria-label-off=\\"off\\"
  data-aria-label-on=\\"on\\"
>
  <img
    alt=\\"visibility_off\\"
    class=\\"mdc-icon-button__icon\\"
    src=\\"./img/visibility_off.png\\"
  />
  <img
    alt=\\"visibility\\"
    class=\\"mdc-icon-button__icon mdc-icon-button__icon--on bar\\"
    src=\\"./img/visibility.png\\"
  />
</button>"
`;

exports[`IconToggle component supports material icon toggle buttons 1`] = `
"<button
  aria-label=\\"visibility\\"
  aria-pressed=\\"false\\"
  class=\\"mdc-icon-button mdc-ripple-upgraded--unbounded\\"
>
  <i
    class=\\"mdc-icon-button__icon material-icons\\"
  >
    visibility_off
  </i>
  <i
    class=\\"mdc-icon-button__icon mdc-icon-button__icon--on material-icons\\"
  >
    visibility
  </i>
</button>"
`;

exports[`IconToggle component supports material icon toggle buttons 2`] = `
"<button
  aria-label=\\"off\\"
  class=\\"mdc-icon-button mdc-ripple-upgraded--unbounded\\"
  data-aria-label-off=\\"off\\"
  data-aria-label-on=\\"on\\"
>
  <i
    class=\\"mdc-icon-button__icon material-icons\\"
  >
    visibility_off
  </i>
  <i
    class=\\"mdc-icon-button__icon mdc-icon-button__icon--on material-icons\\"
  >
    visibility
  </i>
</button>"
`;

exports[`IconToggle component supports material icon toggle buttons 3`] = `
"<button
  aria-label=\\"visibility\\"
  aria-pressed=\\"true\\"
  class=\\"mdc-icon-button mdc-icon-button--on foo mdc-ripple-upgraded--unbounded\\"
>
  <i
    class=\\"mdc-icon-button__icon material-icons\\"
  >
    visibility_off
  </i>
  <i
    class=\\"mdc-icon-button__icon mdc-icon-button__icon--on material-icons\\"
  >
    visibility
  </i>
</button>"
`;
